Feature 23 scatter

bikegeek opened the pull request:
## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>
   Ran the test using the MPR data used for the Land project

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

-    verify that all tests pass
-    proof-read User's Guide to ensure instructions are clear

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes ]**

- [x] Do these changes include sufficient testing updates? **[Yes]**

- [x] Will this PR result in changes to the test suite? **[No]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

- [x] Do these changes introduce new SonarQube findings? **[Yes]**</br>
If **yes**, please describe: Gatecheck flags duplicated code threshold on new code exceeds 2% but older code is included in this metric.  

- [x] Please complete this pull request review by **RC1**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Add any new Python packages to the [METplus Components Python Requirements](https://metplus.readthedocs.io/en/develop/Users_Guide/appendixA.html#metplus-components-python-packages) table.
- [x] Review the source issue metadata (required labels, projects, and milestone).
- [ ] Complete the PR definition above.
- [ ] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the version that will include these changes
Select: **Coordinated METplus-X.Y Support** project for bugfix releases or **METplotpy-X.Y.Z Development** project for official releases
- [x] After submitting the PR, select the :gear: icon in the **Development** section of the right hand sidebar. Search for the issue that this PR will close and select it, if it is not already selected.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

bikegeek committed:
Refactored versions of scatter plot code

bikegeek committed:
replacing with refactored scatter defaults

bikegeek committed:
#23 updates to default values

bikegeek committed:
#23 scatter plot first revision

bikegeek committed:
#23 updates to include missing/mismatched config values

bikegeek committed:
fix grammar in comment

bikegeek committed:
Added support for trendline (regression), caption, cleaned up logging

bikegeek committed:
clean up, adding missing config values, trendline support

bikegeek committed:
Files for testing the scatter plot

bikegeek committed:
Add test for scatter plotting

bikegeek committed:
Change import to see if GHA test passes

bikegeek committed:
try full path for import

bikegeek committed:
change order of tests that are run

bikegeek committed:
revert import of scatter module

bikegeek committed:
needed for pytest to work

bikegeek committed:
Fixed run command

bikegeek committed:
Update and rename reformatted_usecase_data_for_scatter.data to reformatted_data_for_scatter.data

bikegeek committed:
Update test_scatter_mpr.yaml

changed name of data file

bikegeek committed:
updated for change to input data filename

bikegeek committed:
Add check for empty input dataframe, raising error and logging the error

bikegeek committed:
Merge branch 'feature_23_scatter' of https://github.com/dtcenter/METplotpy into feature_23_scatter

bikegeek committed:
Delete metplotpy/plots/scatter/scatter1_plot_data.txt

No longer compatible with refactored code

bikegeek committed:
Delete metplotpy/plots/scatter/scatter2_plot_data.txt

no longer compatible with refactored code

bikegeek committed:
Documentation updates for the scatter plot.

bikegeek committed:
Expected scatter plot from the custom config settings

bikegeek committed:
Merge branch 'feature_23_scatter' of https://github.com/dtcenter/METplotpy into feature_23_scatter

bikegeek committed:
Further refinement of instructions: remove redundant text.

bikegeek committed:
Fixed name of output plot

bikegeek committed:
fixed directory path for output plot

bikegeek committed:
Remove extra * in bold text

bikegeek committed:
Delete metplotpy/plots/scatter/custom_scatter.yaml

no longer relevant to the refactored scatter plot

bikegeek committed:
Delete metplotpy/plots/scatter/scatter_defaults.yml

no longer relevant for the refactored scatter plot

bikegeek committed:
Merge branch 'develop' of https://github.com/dtcenter/METplotpy into feature_23_scatter

bikegeek committed:
Removed out of date README

bikegeek committed:
Merge branch 'develop' of https://github.com/dtcenter/METplotpy into feature_23_scatter

bikegeek committed:
No longer relevant

bikegeek committed:
No longer relevant. New version of scatter plot is accompanied by testing code.

bikegeek committed:
Clean up generated plot

bikegeek committed:
Address SonarQube code smells

bikegeek committed:
Address SonarQube duplicate code warning

bikegeek committed:
Added default values that are needed for METviewer, even though this plot is not used by METviewer.
This plot now works with modifications to the config.py base class that address code duplication in SonarQube

bikegeek committed:
Removed duplicate code and moved to base configuration file config.py

bikegeek committed:
Addresses SonarQube finding of duplicate code.  Duplicate code is now encapsulated in this base configuration class.

bikegeek committed:
Added defaults needed by base config (and for future incorporation to METviewer)

bikegeek committed:
Added mar_offset setting True/False to allow config base class to determine whether to apply an offset to the margin values.  This applies to most of the plots implemented in Plotly.

bikegeek committed:
Added logic to determine whether to apply offset values to margins

bikegeek committed:
Removed duplicated config code that is in the base config class

bikegeek committed:
Removed duplicate code that is now located in the base configuration class

bikegeek committed:
If plot units are already in inches, then use the existing height and width, completing the if-statement that checks if the user has incorrectly set the units to anything other than inches

bikegeek committed:
Needed to add the mar_offset: True to apply the margin offset for this Plotly plot

bikegeek committed:
Requires offset for the margins

bikegeek committed:
Removed remaining duplicated code

bikegeek committed:
Removing some code smells.

jprestop commented on docs/Users_Guide/scatter.rst:
```diff
@@ -0,0 +1,227 @@
+Description
+===========
+The scatter plot is useful for illustrating relationships between pairs of continuous variables.
+This plot was developed to support plotting MPR (matched pair) data from the MET point-stat tool.
```
```suggestion
This plot was developed to support plotting MPR (matched pair) data from the MET Point-Stat tool.
```
Suggesting capitalization to match MET documentation

jprestop commented on docs/Users_Guide/scatter.rst:
```diff
@@ -0,0 +1,227 @@
+===========
+The scatter plot is useful for illustrating relationships between pairs of continuous variables.
+This plot was developed to support plotting MPR (matched pair) data from the MET point-stat tool.
+**NOTE:** This MET output data must first be reformatted into a format that can be read in by the scatter plot code.
```
Suggest using the note directive (.. note:) to be consistent with other notes in the documentation.

bikegeek replied:
done

jprestop commented on docs/Users_Guide/scatter.rst:
```diff
@@ -0,0 +1,227 @@
+**NOTE:** This MET output data must first be reformatted into a format that can be read in by the scatter plot code.
+This reformatting was accomplished through the METdataio METreformat module. The reformatted data
+consists solely of MPR linetype data and all the column headers are labelled according to the
+MPR linetype column names specified in the MET User's Guide, Table 11.20 (Point-stat tool).
```
```suggestion
MPR linetype column names specified in the Point-Stat section of the MET User's Guide in `Table 11.20 <https://metplus.readthedocs.io/projects/met/en/latest/Users_Guide/point-stat.html#table-ps-format-info-seeps>`_ "Format information for SEEPS (Stable Equitable Error in Probability Space) output line type".
```
Suggest adding a link and more information, like the table name, in case the link should break, which hopefully it wouldn't, but just in case.

jprestop commented on docs/Users_Guide/scatter.rst:
```diff
@@ -0,0 +1,227 @@
+Sample Data
+-----------
+
+The data is text output from the MET point-stat tool. The data is arranged in columnar format and reformatted by the
```
```suggestion
The data is text output from the MET Point-Stat tool. The data is arranged in columnar format and reformatted by the
```

jprestop commented on docs/Users_Guide/scatter.rst:
```diff
@@ -0,0 +1,227 @@
+The sample data for creating the example scatter plot is available in the
+`METplotpy <https://github.com/dtcenter/METplotpy>`_ repository, where the scatter plot tests are located:
+
+    *$METPLOTPY_BASE/metplotpy/test/scatter/reformatted_data_for_scatter.data*
```
```suggestion
    *$METPLOTPY_BASE/test/scatter/reformatted_data_for_scatter.data*
```
It looks like test is at the same level as metplotpy and not a subdir under metplotpy.

jprestop commented on docs/Users_Guide/scatter.rst:
```diff
@@ -0,0 +1,227 @@
+*/usr/path/to/METplotpy-x.y.z* if the source code was downloaded as a zip or gzip'd tar file from the Release link of
+the Github repository.  The *x.y.z* is the release number.
+
+**NOTE**:
```
Suggest using the note directive (.. note:) to be consistent with other notes in the documentation.

bikegeek replied:
done

jprestop commented on docs/Users_Guide/scatter.rst:
```diff
@@ -0,0 +1,227 @@
+the Github repository.  The *x.y.z* is the release number.
+
+**NOTE**:
+The MPR linetype data produced by the MET point-stat tool has been reformatted via the METdataio METreformat
```
```suggestion
The MPR linetype data produced by the MET Point-Stat tool has been reformatted via the METdataio METreformat
```

jprestop reviewed (changes requested):
Nice work removing those code smells, @bikegeek.  I have verified that all tests pass and proof-read User's Guide to ensure instructions are clear, without stepping through the process.  I made some minor suggestions for formatting.  Please let me know if you have any questions.

jprestop commented on docs/Users_Guide/scatter.rst:
```diff
@@ -0,0 +1,227 @@
+**DOES NOT require any modifications**.
+
+The second required configuration file is a user-supplied “custom”
+configuration file. This file is used to customize/over ride the default
```
```suggestion
configuration file. This file is used to customize/override the default
```

jprestop commented on docs/Users_Guide/scatter.rst:
```diff
@@ -0,0 +1,227 @@
+The second required configuration file is a user-supplied “custom”
+configuration file. This file is used to customize/over ride the default
+settings in the **scatter_defaults.yaml** file. The custom configuration
+file can contain only those settings that will over ride the default settings in the
```
```suggestion
file can contain only those settings that will override the default settings in the
```

bikegeek committed:
Update docs/Users_Guide/scatter.rst

Co-authored-by: Julie Prestopnik <jpresto@ucar.edu>

bikegeek committed:
Update docs/Users_Guide/scatter.rst

Co-authored-by: Julie Prestopnik <jpresto@ucar.edu>

bikegeek committed:
Update docs/Users_Guide/scatter.rst

Co-authored-by: Julie Prestopnik <jpresto@ucar.edu>

bikegeek committed:
Update docs/Users_Guide/scatter.rst

Co-authored-by: Julie Prestopnik <jpresto@ucar.edu>

bikegeek committed:
Update docs/Users_Guide/scatter.rst

Co-authored-by: Julie Prestopnik <jpresto@ucar.edu>

bikegeek committed:
Update docs/Users_Guide/scatter.rst

Co-authored-by: Julie Prestopnik <jpresto@ucar.edu>

bikegeek committed:
Update docs/Users_Guide/scatter.rst

Co-authored-by: Julie Prestopnik <jpresto@ucar.edu>

bikegeek committed:
Replace bold NOTE instances with RST ' .. note:: ' insted

bikegeek committed:
Address conflicts
Merge branch 'feature_23_scatter' of https://github.com/dtcenter/METplotpy into feature_23_scatter

bikegeek committed:
Found some more point-stat instances that needed to be changed to Point-Stat

jprestop committed:
Added link to table and made "over ride" one word

jprestop reviewed (approved):
Thank you for all of your work on this @bikegeek.  I approve this request.
